test/translate_ligo_library.coffee

```diff
@@ -186,4 +186,158 @@ describe "translate ligo section library", ()->
+        const sum : nat = 0n;
+        sum := (test_reserved_long___self + other);
+        assert((sum >= test_reserved_long___self));
+      } with (sum);
+    (* UsingForDirective *)
+    
+    function uintExactAddOverflowExample (const #{config.reserved}__unit : unit) : (unit) is
+      block {
+        const n : nat = abs(not (0));
+        const terminate_tmp_0 : (nat) = exactMath_exactAdd(n, 1n);
+      } with (unit);
+    """#"
+    make_test text_i, text_o
+  
+  it "library (using) + pure", ()->
+    text_i = """
+    pragma solidity ^0.4.22;
+    
+    library ExactMath {
+      function exactAdd(uint self, uint other) internal pure returns (uint sum) {
+        sum = self + other;
+        require(sum >= self);
+      }
+    }
+    
+    contract Pure_test {
+      using ExactMath for uint;
+      function test() public pure returns (uint) {
+        var n = uint(~0);
+        n.exactAdd(1);
+        return 0;
+      }
+    }
+    """
+    text_o = """
+    type test_args is record
+      callbackAddress : address;
+    end;
+    
+    type state is unit;
+    
+    function exactMath_exactAdd (const test_reserved_long___self : nat; const other : nat) : (nat) is
+      block {
+        const sum : nat = 0n;
+        sum := (test_reserved_long___self + other);
+        assert((sum >= test_reserved_long___self));
+      } with (sum);
+    type router_enum is
+      | Test of test_args;
+    
+    (* UsingForDirective *)
+    
+    function test (const #{config.reserved}__unit : unit) : (nat) is
+      block {
+        const n : nat = abs(not (0));
+        const terminate_tmp_0 : (nat) = exactMath_exactAdd(n, 1n);
+      } with (0n);
+    
+    function main (const action : router_enum; const #{config.contract_storage} : state) : (list(operation) * state) is
+      (case action of
+      | Test(match_action) -> block {
+        const tmp : (nat) = test(unit);
+        var opList : list(operation) := list transaction((tmp), 0mutez, (get_contract(match_action.callbackAddress) : contract(nat))) end;
+      } with ((opList, self))
+      end);
+    """#"
+    make_test text_i, text_o, router: true
   
+  it "using changes storage", ()->
+    text_i = """
+    pragma solidity ^0.5.11;
+    
+    library Bits {
+    
+        uint constant internal const_ONE = uint(1);
+        uint constant internal const_ONES = uint(~0);
+    
+        // Sets the bit at the given 'index' in 'tmp_self' to '1'.
+        // Returns the modified value.
+        function setBit(uint tmp_self, uint8 index) internal pure returns (uint) {
+            return tmp_self | const_ONE << index;
+        }
+    }
+    
+    contract BitsExamples {
+      using Bits for uint;
+    
+      // Set bits
+      function setBitExample() public pure {
+        uint n = 0;
+        n = n.setBit(0); // Set the 0th bit.
+        assert(n == 1);  // 1
+      }
+    }
+    """
+    text_o = """
+    type setBitExample_args is unit;
+    type state is record
+      const_ONE : nat;
```

Comment:
I would suggest translating `constant internal` as global contract-scope constants, not as state variables.
But that is unrelated issue though.